src/main/java/org/zalando/nakadi/repository/TopicRepositoryHolder.java

```diff
@@ -0,0 +1,100 @@
+    private final UUIDGenerator uuidGenerator;
+    private final MetricRegistry metricRegistry;
+    private final ReadWriteLock readWriteLock;
+    private final Map<Storage, TopicRepository> storageTopicRepository;
+
+    @Autowired
+    public TopicRepositoryHolder(final NakadiSettings nakadiSettings,
+                                 final KafkaSettings kafkaSettings,
+                                 final ZookeeperSettings zookeeperSettings,
+                                 final UUIDGenerator uuidGenerator,
+                                 final MetricRegistry metricRegistry) {
+        this.nakadiSettings = nakadiSettings;
+        this.kafkaSettings = kafkaSettings;
+        this.zookeeperSettings = zookeeperSettings;
+        this.uuidGenerator = uuidGenerator;
+        this.metricRegistry = metricRegistry;
+        this.readWriteLock = new ReentrantReadWriteLock();
+        this.storageTopicRepository = new HashMap<>();
+    }
+
+    public TopicRepository getTopicRepository(final Storage storage) throws TopicRepositoryException {
+        readWriteLock.readLock().lock();
+        try {
+            final TopicRepository topicRepository = storageTopicRepository.get(storage);
+            if (topicRepository != null) {
+                return topicRepository;
+            }
+        } finally {
+            readWriteLock.readLock().unlock();
+        }
+
+        readWriteLock.writeLock().lock();
+        try {
+            TopicRepository topicRepository = storageTopicRepository.get(storage);
+            if (topicRepository != null) {
+                return topicRepository;
+            }
+            topicRepository = createTopicRepository(storage);
+            storageTopicRepository.put(storage, topicRepository);
+            return topicRepository;
+        } finally {
+            readWriteLock.writeLock().unlock();
+        }
+    }
+
+    private TopicRepository createTopicRepository(final Storage storage) throws TopicRepositoryException {
+        switch (storage.getType()) {
+            case KAFKA:
+                try {
+                    final Storage.KafkaConfiguration kafkaConfiguration = storage.getKafkaConfiguration();
+                    final ZooKeeperHolder zooKeeperHolder = new ZooKeeperHolder(
+                            kafkaConfiguration.getZkAddress(),
+                            kafkaConfiguration.getZkPath(),
+                            kafkaConfiguration.getExhibitorAddress(),
+                            kafkaConfiguration.getExhibitorPort());
+                    final KafkaFactory kafkaFactory =
+                            new KafkaFactory(new KafkaLocationManager(zooKeeperHolder, kafkaSettings), metricRegistry);
+                    final KafkaTopicRepository kafkaTopicRepository = new KafkaTopicRepository(zooKeeperHolder,
+                            kafkaFactory, nakadiSettings, kafkaSettings, zookeeperSettings, uuidGenerator);
+                    return kafkaTopicRepository;
```

Comment:
maybe it makes sense to check if it is possible to call at least something from this topic repository (like list topics)